### PR TITLE
feature: #100 Distribute Goxygen as a Docker image

### PR DESCRIPTION
**Pull request by orgmatileg**
Hi I just made changes for #100 distributing Goxygen to GitHub Packages Docker

What changes will happen:
- When the branch master has a new commit, GitHub Action `docker-build-package` will trigger and create an image and push as `ghcr.io/Shpota/goxygen:latest`
- When a new tag is created GitHub Action `docker-build-package` will trigger and create an image and push as `ghcr.io/Shpota/goxygen:{new tag name}`

### CODE REVIEW

**Comment by orgmatileg** [comment]
Hi, Gentle reminder to review the PR.. thanks!

Have a nice day

**Comment by Shpota** [comment]
I am sorry for the delay @orgmatileg, I am quite busy these days. I would need to try out, run, and foremost to understand your code, which requires time. I'll review it once I find time (hopefully this week).

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM golang:1.17.6-bullseye
```
Comment (Shpota) [inline Dockerfile:1]: 1. I would suggest using an alpine-based image instead (with git on top of it). The bullseye image takes up almost 1GB in comparison with 329Mb of an alpine one.
2. Let's use a multistage docker build. After a Goxygen binary is built, we don't need its source code in the image as well as the Go compiler, and executor. You can copy the `/app/goxygen` to the new image (an alpine-based one). In the end, the image would be only 24Mb in size.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ENV GOXYGEN_DOCKER=true
+
+WORKDIR /app
+COPY . .
+
+# Without this line, the docker image will not be able to run and produce error: 
+# `Failed to setup a Git repository: exit status 128`
+# Because there is a line of code that stage files to the git repository and commit on
+# codegen > codegen.go > function initGitRepo 
+RUN git config --global user.email "arlhba@gmail.com" && git config --global user.name "Luqmanul Hakim"
```
Comment (Shpota) [inline Dockerfile:14]: I would suggest using a non-existing email for code generation. Eg: `goxygen@noreply.github.com` with name set as `Goxygen`.

If you want to preserve attribution in the image, you can add a label that would end up in the image metadata once it is pushed:
```
LABEL org.opencontainers.image.authors="arlhba@gmail.com"
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+RUN go build -o goxygen
+
+
+
```
Comment (Shpota) [inline Dockerfile:19]: I would suggest relapsing three empty lines with one.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
+FROM golang:1.17.6-bullseye
+
+ENV UID=1000
+ENV GID=1000
+ENV GOXYGEN_DOCKER=true
```
Comment (Shpota) [inline Dockerfile:5]: I don't quite understand why this part is needed (as well as the corresponding changes in the Go code).

I have tested and the following command would correctly do what we need (given there is a symlink to goxygen executable and the workdir of an image is set to `/app`):
```
docker run --rm -it -v $(pwd):/app image-tag goxygen init my-app
```

Let's optimize it.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+WORKDIR /app
+COPY . .
+
+# Without this line, the docker image will not be able to run and produce error:
```
Comment (Shpota) [inline Dockerfile:10]: Let's remove these comments. I guess it is fine considering the app is supposed to generate a Git repository.

File: .github/workflows/docker-build-package.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Build image
+        run: sudo docker build . --file Dockerfile --tag $IMAGE_NAME --label "runnumber=${GITHUB_RUN_ID}"
+
+      - name: Log in to registry
+        # This is where you will update the PAT to GITHUB_TOKEN
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
+
+      - name: Push image
```
Comment (Shpota) [inline .github/workflows/docker-build-package.yml:31]: I didn't test this part yet. Can you please tell me what tags would it produce and for what conditions? I want to make sure I understand the code correctly.

**Review by Shpota — CHANGES_REQUESTED** [review]
Thank you for submitting this pull request!
I am sorry for not reviewing it for so long. With the war in my country, I didn't have time for open source. I added a couple of comments suggesting improvements. Let me know what you think.

**Comment by Shpota** [comment]
Thank you for the pull request.

For go 1.17+ the user can run `go run package@version` syntax which removes the need of having a Docker image for generating a project. I believe it is sufficient.

Closing this PR.

**Pull request closed without merging**